### ISSUE TITLE
Feature Request: Support for YARN cluster manager for Distributed TensorFlow

### ISSUE THREAD

**Issue by j9ac9k** [tensorflow/tensorflow#2655]
The Distributed TensorFlow documentation said to open an issue to request support for a specific cluster manager, support for YARN would be beneficial. 

Thanks

**Comment by martinwicke** [comment]
Yes, we'd like to do that. Nobody is working on it right now though.

**Comment by bhack** [comment]
@j9ac9k In the meantime if you are interested there is an initial work on Mesos that already can use GPU resources at https://github.com/tensorflow/tensorflow/issues/1996

**Comment by zjujimmy** [comment]
Is anyone working on this now?

**Comment by mrry** [comment]
We've started looking at this internally. @jhseu will be leading up this effort.

**Comment by bhack** [comment]
So are you supporting internally yarn and on contribution mesos?

**Comment by jhseu** [comment]
We'll likely do both.

**Comment by jimdowling** [comment]
Great. Any timeline on this?

**Comment by jhseu** [comment]
There are some prerequisites that need to be done first for it to be useful (e.g. HDFS support). Definitely soonish, but I'd prefer not commit externally to a timeline :)

**Comment by jhseu** [comment]
Also, YARN has some complications that don't exist in Mesos and Kubernetes, so we're not completely sure we'll add support for it.

**Comment by dzhwinter** [comment]
How tensorflow will support hdfs ? Any progress? I'm going to contribute on it.

**Comment by martinwicke** [comment]
We're working on HDFS, it's close.

**Comment by niuchenglei** [comment]
@martinwicke hi guy, how does the HDFS support going on? very looking forward to this feature.

**Comment by nchivukula** [comment]
@martinwicke Looking forward to this feature. Let me know if you need any help on this to get it out quicker

**Comment by JunpingDu** [comment]
Hi @jhseu and @martinwicke , I am from YARN community. Just noticed some comments above to say extra complications on YARN over Mesos/Kubernetes. Can you shed more light on this? I can see how can I help here, especially from YARN side. Thx!

**Comment by butterluo** [comment]
Any update ??

**Comment by jhseu** [comment]
No update. We'll publish Mesos/Kubernetes examples fairly soon and then consider YARN next.

@JunpingDu I'll follow up with you through e-mail.

**Comment by dzhwinter** [comment]
is there any HDFS support release plan recently? 
Thanks.

**Comment by martinwicke** [comment]
HDFS support is available in 0.11 (and at head). See this [howto](https://www.tensorflow.org/versions/r0.11/how_tos/hadoop/index.html).

**Comment by jhseu** [comment]
You can ignore those instructions and just start using hdfs paths if you use this Docker image:
https://github.com/tensorflow/ecosystem/blob/master/docker/Dockerfile.hdfs

**Comment by wangyang0918** [comment]
I am working on this and will use Slider to support tensorflow on YARN. I have created a JIRA in Slider Project to make a introduction. https://issues.apache.org/jira/browse/SLIDER-1174

**Comment by jhseu** [comment]
Nice! When you have a working configuration, would you mind sending a pull request to [github.com/tensorflow/ecosystem](https://github.com/tensorflow/ecosystem)?

**Comment by wangyang0918** [comment]
Sure, I will send a pull request to ecosystem when it could work

**Comment by tangzhankun** [comment]
@JunpingDu @jhseu , this is zhankun from intel focused on Hadoop. We are glad to help on YARN too. How can I join your thread? :)

**Comment by formath** [comment]
@wangyang0918 A great job. Any progresses?

**Comment by wangyang0918** [comment]
Thanks for your attention. @formath 
I have already done most of the INCLUDED features in SLIDER-1174 and will attach a alpha patch next week.

**Comment by c-val** [comment]
Hi! Yarn support would be great! Could you share your progress please? Thank you for your efforts!

**Comment by drpngx** [comment]
@wangyang0918 I didn't find a PR there yet. Hope you're progressing.

**Comment by manuzhang** [comment]
@drpngx Note there is also [YARN-6043 TensorFlow on YARN](https://issues.apache.org/jira/browse/YARN-6043) in the Hadoop community that doesn't require docker. You may check out the details in the design document.

**Comment by zhe-thoughts** [comment]
I have a high level question regarding this feature request: what are the key benefits of a native approach (directly running TF on YARN) compared with a Spark based approach, most notably https://github.com/yahoo/TensorFlowOnSpark?

**Comment by leftnoteasy** [comment]
In Hadoop 3.0, YARN native services can support running Tensorflow services on YARN without adding any dependencies or implement a new YARN application master.

Please see our blogpost: https://hortonworks.com/blog/distributed-tensorflow-assembly-hadoop-yarn/ and let me know if you have any questions. Thanks!

**Comment by drpngx** [comment]
Nice

**Comment by tobiajo** [comment]
In the meantime, while waiting for Hadoop 3.0, this could be an alternative, an YARN native implementation:
https://github.com/hopshadoop/hops-tensorflow

**Comment by xiaoyongzhu** [comment]
Not sure there's any update for this effort?

**Comment by jimdowling** [comment]
Just to echo @tobiajo  - there is a version of distributed YARN-native tensorflow available here:
https://github.com/hopshadoop/hops-tensorflow
This also includes support for GPUs as a resource in YARN:
https://hopshadoop.atlassian.net/browse/HOPS-11

GUI support for running distributed TF jobs on YARN will be provided in Hopsworks by the end of the month (including Jupyter to write/run jobs):
https://www.youtube.com/watch?v=rop6bhuBHT0

**Comment by xiaoyongzhu** [comment]
@jimdowling Thanks for the pointers. One quick question - does this YARN-native Tensorflow work with regular Hadoop (i.e. Apache Hadoop, rather than Hops)?

**Comment by jimdowling** [comment]
@xiaoyaozhuzi  - no. Hops Hadoop is a fork of Apache Hadoop. The YARN code is identical to the Apache code (version 2.7.3) with only a few additions: cluster state is recovered from MySQL Cluster and support for GPUs. If somebody wanted they could take our code and introduce it to YARN. But the YARN folks would prob reject the PR as they want to redesign YARN's resource model before accepting support for GPUs. That will take time.

**Comment by tobiajo** [comment]
@xiaoyaozhuzi no since we added GPU support as a resource Apache Hadoop is not supported. One could, with just a little effort, fork the repo, strip the GPU stuff and just change the dependencies in the [pom.xml](https://github.com/hopshadoop/hops-tensorflow/blob/master/pom.xml) to Apache Hadoop.

**Comment by tensorflowbutler** [comment]
It has been 14 days with no activity and this issue has an assignee.Please update the label and/or status accordingly.

**Comment by martinwicke** [comment]
I'm closing this -- a feature request for YARN support and any discussion should probably happen at  tensorflow/ecosystem instead.
